### ISSUE TITLE
csrc/hetero_reduce: cross_entropy_tp — heterogeneous tensor-parallel loss with PCIe-aware vocab partition

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#141]
## Problem
cross_entropy_tp.cu assumes uniform vocab partition across TP ranks. On our heterogeneous mesh:
- H100 (SM9.0) can hold 2x the vocab slice of A6000 (SM8.6)
- Current uniform partition wastes H100 memory and underutilizes A6000

### Current
```cpp
void cross_entropy_tp_forward(
    const half* logits,  // [B*S, V/TP]  uniform partition
    const int* targets,  // [B*S]
    float* loss,         // [B*S]
    int B_S, int V_per_rank, int tp_rank, int tp_size,
    cudaStream_t stream);
```

### Target
```cpp
struct VocabPartition {
    int offset;   // this ranks vocab start
    int count;    // this ranks vocab size (non-uniform)
};

void cross_entropy_tp_forward(
    const half* logits,           // [B*S, local_V]
    const int* targets,           // [B*S]
    float* loss, float* max_logit, float* sum_exp,
    int B_S, VocabPartition part,
    cudaStream_t stream);

// Two-pass: local max/sum → allreduce → local loss
// Allreduce uses pcie_adaptive_allreduce (#138) for PCIe mesh
```

## Language: C++/CUDA
